tools/keyvaluestore.go

```diff
@@ -0,0 +1,201 @@
+// only locked during load and save, to make it concurrency friendly. When
+// saving, the store uses optimistic locking to determine whether the db on disk
+// has been modified by another process; in which case it loads the latest
+// version and re-applies modifications made during this session. This means
+// the Lost Update db concurrency issue is possible; so don't use this if you
+// need more DB integrity than Read Committed isolation levels.
+type KeyValueStore struct {
+	mu       sync.RWMutex
+	filename string
+	log      []keyValueChange
+
+	// This is the persistent data
+	// version for optimistic locking, this field is incremented with every Save()
+	version int64
+	db      map[string]interface{}
+}
+
+type keyValueOperation int
+
+const (
+	// Set a value for a key
+	keyValueSetOperation = keyValueOperation(iota)
+	// Removed a value for a key
+	keyValueRemoveOperation = keyValueOperation(iota)
+)
+
+type keyValueChange struct {
+	operation keyValueOperation
+	key       string
+	value     interface{}
+}
+
+// NewKeyValueStore creates a new store and initialises it with contents from
+// the named file, if it exists
+func NewKeyValueStore(filepath string) (*KeyValueStore, error) {
+	kv := &KeyValueStore{filename: filepath, db: make(map[string]interface{})}
+	err := kv.loadAndMergeIfNeeded()
+	return kv, err
+}
+
+// Set updates the key/value store in memory
+// Changes are not persisted until you call Save()
+func (k *KeyValueStore) Set(key string, value interface{}) {
+	k.mu.Lock()
+	defer k.mu.Unlock()
+
+	k.db[key] = value
+	k.log = append(k.log, keyValueChange{keyValueSetOperation, key, value})
+}
+
+// Remove removes the key and its value from the store in memory
+// Changes are not persisted until you call Save()
+func (k *KeyValueStore) Remove(key string) {
+	k.mu.Lock()
+	defer k.mu.Unlock()
+
+	delete(k.db, key)
+	k.log = append(k.log, keyValueChange{keyValueRemoveOperation, key, nil})
+}
+
+// Get retrieves a value from the store, or nil if it is not present
+func (k *KeyValueStore) Get(key string) interface{} {
+	// Read-only lock
+	k.mu.RLock()
+	defer k.mu.RUnlock()
+
+	// zero value of interface{} is nil so this does what we want
+	return k.db[key]
+}
+
+// Save persists the changes made to disk
+// If any changes have been written by other code they will be merged
+func (k *KeyValueStore) Save() error {
+	k.mu.Lock()
+	defer k.mu.Unlock()
+
+	// Short-circuit if we have no changes
+	if len(k.log) == 0 {
+		return nil
+	}
+
+	// firstly peek at version; open read/write to keep lock between check & write
+	f, err := os.OpenFile(k.filename, os.O_RDWR|os.O_CREATE, 0664)
+	if err != nil {
+		return err
+	}
+	stat, _ := os.Stat(k.filename)
+
+	defer f.Close()
+
+	// Only try to merge if > 0 bytes, ignore empty files (decoder will fail)
+	if stat.Size() > 0 {
+		k.loadAndMergeReaderIfNeeded(f)
+		// Now we overwrite the file
+		f.Seek(0, os.SEEK_SET)
+		f.Truncate(0)
+	}
+
+	k.version++
+
+	enc := gob.NewEncoder(f)
```

Comment:
Actually, we may want to just hold onto this instance instead of the file. Perhaps both, since we need to `f.Seek` and `f.Truncate`.